klient/k8s/watcher/watch.go

```diff
@@ -0,0 +1,112 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package watcher
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/rest"
+	cr "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/e2e-framework/klient/k8s"
+)
+
+// EventHandlerFuncs is an adaptor to let you easily specify as many or
+// as few of functions to invoke while getting notification from watcher
+type EventHandlerFuncs struct {
+	addFunc     func(obj interface{})
+	updateFunc  func(newObj interface{})
+	deleteFunc  func(obj interface{})
+	watcher     watch.Interface
+	ListOptions *cr.ListOptions
+	K8sObject   k8s.ObjectList
+	Cfg         *rest.Config
+}
+
+// EventHandler can handle notifications for events that happen to a resource.
+// Start will be waiting for the events notification which is responsible
+// for invoking the registered user defined functions.
+// Stop used to stop the watcher.
+type EventHandler interface {
+	Start(ctx context.Context)
+	Stop()
+}
+
+// Start triggers the registered methods based on the event received for
+// particular k8s resources
+func (e *EventHandlerFuncs) Start(ctx context.Context) error {
+	cl, err := cr.NewWithWatch(e.Cfg, cr.Options{})
+	if err != nil {
+		return err
+	}
+
+	w, err := cl.Watch(ctx, e.K8sObject, e.ListOptions)
+	if err != nil {
+		return err
+	}
+
+	// set watcher object
+	e.watcher = w
+
+	for event := range e.watcher.ResultChan() {
```

Comment:
2) here the loop needs to check to make sure the context has not been cancelled. One way to do this is to use a for-select loop block as shown.

Also, note I am suggesting that the event processing loop be wrapped in a goroutine.  I think it's better if Start automatically does that rather than the caller of Start.

```go
func (e *EventHandlerFuncs) Start(ctx context.Context) error {

    go func() {
        for {
            select {
                case <-ctx.Done() :
                    if ctx.Err() != nil {
                        return ctx.Err()
                    }
                    return nil
                case event := <- e.watcher.ResultChan():
                    // process event here
            }
        }
    }()

     return nil
...
}